### PR TITLE
Add support for init_meta_context, materialize_module

### LINKED ISSUE #9375: [RFC] Simplify sharding API instantiation
## 🚀 Feature

Currently, the Lightning users working with sharded models and with DeepSpeed or FSDP Plugin uses condi 

Here is an example.

```py
    class MyModel(pl.LightningModule):
        def __init__(self):
            super().__init__()
            self.linear_layer = nn.Linear(32, 32)
            self.block = nn.Sequential(nn.Linear(32, 32), nn.ReLU())
            self.final_block = nn.Sequential(nn.Linear(32, 32), nn.ReLU())

        def configure_sharded_model(self):
            # modules are sharded across processes
            # as soon as they are wrapped with ``wrap`` or ``auto_wrap``.
            # During the forward/backward passes, weights get synced across processes
            # and de-allocated once computation is complete, saving memory.

            # Wraps the layer in a Fully Sharded Wrapper automatically
            linear_layer = wrap(self.linear_layer)

            # Wraps the module recursively
            # based on a minimum number of parameters (default 100M parameters)
            block = auto_wrap(self.block)

            # For best memory efficiency,
            # add FairScale activation checkpointing
            final_block = auto_wrap(checkpoint_wrapper(self.final_block))
            self.model = nn.Sequential(linear_layer, nn.ReLU(), block, final_block)

        def configure_optimizers(self):
            return torch.optim.AdamW(self.model.parameters())
```

It could be possibly to uniformize the API using the new meta device introduced by PyTorch 1.9.0: https://pytorch.org/tutorials/prototype/skip_param_init.html

```py
from pytorch_lightning.distributed import skip_param_init, apply_param_init

class Model(LightningModule):

    def __init__(self):
        super()

        with skip_param_init():
            # create parameters as before, possibly ones which don't fit on 1 gpu


    def setup(self, ...):
        ...


# This model doesn't have any params for module defined within skip_param_init context manager
model = Model() #

# In DeepSpeed / FSDP, distributed is fully available, `apply_param_init` is applied to create the sharded weights.
```

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #9375

This PR builds on top of https://github.com/pytorch/pytorch/pull/66317. A code section will be dropped once merged to PyTorch.

The goal is no code change for the end users.

```py
from pytorch_lightning.utilities.meta import init_meta_context, materialize_module

class MLP(nn.Module):
    def __init__(self, num_convs: int):
        super().__init__()
        self.lins = []
        for _ in range(num_convs):
            self.lins.append(nn.Linear(1, 1))
        self.layer = nn.Sequential(*self.lins)

with init_meta_context():
    m = nn.Linear(in_features=1, out_features=1)
    assert m.weight.device.type == "meta"
    mlp = MLP(4)
    assert mlp.layer[0].weight.device.type == "meta"

    materialize_module(mlp)
    assert mlp.layer[0].weight.device.type == "cpu"

m = nn.Linear(in_features=1, out_features=1)
assert m.weight.device.type == "cpu"

with init_meta_context():
    m = nn.Linear(in_features=1, out_features=1)
    assert m.weight.device.type == "meta"

m = nn.Linear(in_features=1, out_features=1)
assert m.weight.device.type == "cpu"
```

TODO:
- [x] Verify it works with DeepSpeed on BoringModel

```py
    with init_meta_context():
        model = BoringModel()
    assert model.layer.weight.device.type == "meta"
    trainer = Trainer(
        default_root_dir=tmpdir, plugins=[DeepSpeedPlugin(stage=3)], gpus=2, fast_dev_run=True, precision=16
    )
    trainer.fit(model)
    assert model.layer.weight.device.type == "cpu"
```

- [x] Verify it works with DeepSpeed on minGPT. Find the associated example here: https://github.com/SeanNaren/minGPT/blob/meta_device/train.py

Screenshot 2021-10-15 at 12 59 53

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by SeanNaren** [comment]
This is incredible work allowing users to not have to change their model definition, and hopefully in the future FSDP supports sharding directly from the meta device modules.

(cc @myleott @blefaudeux @anj-s who may have some stuff to say about the API/future integration!)

Should we wait for the code to be merged into PyTorch and available in the nightly build?

**Comment by tchaton** [comment]
> This is incredible work allowing users to not have to change their model definition, and hopefully in the future FSDP supports sharding directly from the meta device modules.
> 
> (cc @myleott @blefaudeux @anj-s who may have some stuff to say about the API/future integration!)
> 
> Should we wait for the code to be merged into PyTorch and available in the nightly build?

Adding @cbalioglu to the conversation.

IMO, we definitely want this for `Lightning v1.5` as it is already working with PyTorch 1.10 nightly. 
Hopefully, this feature will be merged in 1.10 and released before Lightning v1.5, so we can remove the copy / paste.

Best,
T.C

**Comment by anj-s** [comment]
> This is incredible work allowing users to not have to change their model definition, and hopefully in the future FSDP supports sharding directly from the meta device modules.
Yep! +1 on this effort. I generally think we need better solutions for checkpoint loading/saving, model instantiation and initialization. this is something we needed to implement SSD offloading for large models and I had a prototype with a meta device being specified for every layer to help with model initialization. Unfortunately this was not viable because there is no way to change every layer of a model! 

> (cc @myleott @blefaudeux @anj-s who may have some stuff to say about the API/future integration!)
Thanks for pinging! Added comments on the PyTorch PR since I am assuming we will move to it?

> Should we wait for the code to be merged into PyTorch and available in the nightly build?
@tchaton This PR seems to have a different API. Will we use this new API or the one provided by PyTorch?

**Comment by tchaton** [comment]
Hey @ben

> > The goal for the end users is no code change !!!
> 
> Are you referring to no code changes to the model instantiation or no API changes?
> 
> Because if I understood everything, the API this PR proposes and the one upstream are different which will force an eventual update.
> 
> Also, what do we gain with trying to merge this before it's available upstream? Is it just to start debugging and testing integrations? Do we expect any users to start using it at this stage?

Hey @carmocca 

1. I am refering to no code changes to the model instantiation

2. The PR upstream provide the init_meta functionality. This PR builds on it to enable automatic meta device creation without having to change the model definition. It hides the logic behind the _MetaClass __new__ function.

3. I am pretty sure it should work with DeepSpeed, just need to check it out. It means we could already get user feedbacks and possibly demonstrate to PyTorch code to need for such utility.
 
Dear @anj-s @blefaudeux, yes a context manager utility can easily be added. The current API was more related to this desired API in PyTorch:

```py
torch.set_device(torch.device("meta"))  
```

In Lightning, it is likely the model instantiation and materialisation might happen outside the context manager (c.f DeepSpeed Plugin.)

Best,
T.C
